R/msgmerge.R

```diff
@@ -31,28 +35,72 @@ run_msgfmt = function(po_file, mo_file, verbose) {
   return(invisible())
 }
 
-update_mo_files = function(dir, package, verbose) {
-  inst_dir <- file.path(dir, "inst", "po")
-  lang_regex <- "^(R-)?([a-zA-Z_]+)\\.po$"
+#' Compile `.po` files to `.mo`
+#'
+#' This function compiles the plain text `.po` files that translators work with
+#' into the binary `.mo` files that are installed with packages and used for
+#' live translations.
+#'
+#' @param dir Path to package root directory.
+#' @param package Name of package. If not supplied, read from `DESCRIPTION`.
+#' @param lazy If `TRUE`, only `.mo` functions that are older than `.po`
+#'   files be updated
+#' @param verbose If `TRUE`, print information as it goes.
+po_compile = function(dir = ".", package = NULL, lazy = TRUE, verbose = TRUE) {
+  po_metadata <- get_po_metadata(dir = dir, package = package, lazy = lazy)
+  dir_create(dirname(po_metadata$mo))
 
-  po_files <- list.files(file.path(dir, "po"), pattern = "\\.po$")
-  languages <- gsub(lang_regex, "\\2", po_files)
-  mo_files <- gsub(lang_regex, sprintf("\\1%s.mo", package), po_files)
-  mo_dirs <- file.path(inst_dir, languages, "LC_MESSAGES")
-  # NB: dir.create() only accepts one directory at a time...
-  for (mo_dir in unique(mo_dirs)) dir.create(mo_dir, recursive = TRUE, showWarnings = FALSE)
+  if (!lazy) {
+    # Clear out all older translations
+    mo_dirs <- dir(file.path(dir, "inst", "po"), full.names = TRUE)
+    to_delete <- mo_dirs[!basename(mo_dirs) %in% c(po_metadata$language, "en@quot")]
+
+    for (dir in to_delete) {
+      if (verbose) messagef(
+        "Found a compiled translation for %s at %s, but no corresponding .po; deleting",
+        basename(dir), dirname(dir)
+      )
+      unlink(dir, recursive = TRUE)
+    }
+  }
 
-  for (ii in seq_along(po_files)) {
+  for (ii in seq_len(nrow(po_metadata))) {
+    if (verbose) messagef("Recompiling %s translation", po_metadata$language[ii])
     run_msgfmt(
-      po_file = file.path(dir, "po", po_files[ii]),
-      mo_file = file.path(mo_dirs[ii], mo_files[ii]),
+      po_file = po_metadata$po[ii],
+      mo_file = po_metadata$mo[ii],
       verbose = verbose
     )
   }
 
   return(invisible())
 }
 
+get_po_metadata <- function(dir = ".", package = NULL, lazy = TRUE) {
```

Comment:
felt a bit odd that po_files() returns an object involving po & mo files. i think `get_po_metadata` is an improvement but maybe still not ideal, WDYT

Reply:
I was thinking `po_` as the generic suffix for potools, not specifically for `.po` files. `po_metadata()` also works for me; generally I think the `get_` is implied for function names that are nouns, but I don't have strong feelings either way.